sway/tree/view.c

```diff
@@ -683,6 +683,38 @@ void view_child_destroy(struct sway_view_child *child) {
 	}
 }
 
+struct sway_view *view_from_wlr_surface(struct wlr_surface *wlr_surface) {
+again:
+	if (wlr_surface_is_xdg_surface(wlr_surface)) {
+		struct wlr_xdg_surface *xdg_surface =
+			wlr_xdg_surface_from_wlr_surface(wlr_surface);
+		return view_from_wlr_xdg_surface(xdg_surface);
+	}
+	if (wlr_surface_is_xdg_surface_v6(wlr_surface)) {
+		struct wlr_xdg_surface_v6 *xdg_surface_v6 =
+			wlr_xdg_surface_v6_from_wlr_surface(wlr_surface);
+		return view_from_wlr_xdg_surface_v6(xdg_surface_v6);
+	}
+	if (wlr_surface_is_xwayland_surface(wlr_surface)) {
+		struct wlr_xwayland_surface *xsurface =
+			wlr_xwayland_surface_from_wlr_surface(wlr_surface);
+		return view_from_wlr_xwayland_surface(xsurface);
+	}
+	if (wlr_surface_is_subsurface(wlr_surface)) {
+		struct wlr_subsurface *subsurface =
+			wlr_subsurface_from_surface(wlr_surface);
+		wlr_surface = subsurface->parent;
+		goto again;
```

Comment:
What about

```c
return view_from_wlr_surface(subsurface->parent)
```

?

Reply:
I would do that in a language that optimizes tail-recursion... But I guess once again we don't really care, I'll do that.

Reply:
Pretty sure C compilers optimize tail calls at `-O2` or `-O3`

Reply:
hm, right, I wonder why I was convinced it didn't. Guess optimisations weren't turned on properly when I last checked